source/extensions/filters/network/dynamic_modules/abi_impl.cc

```diff
@@ -265,6 +269,270 @@ void envoy_dynamic_module_callback_network_filter_disable_close(
   }
 }
 
+void envoy_dynamic_module_callback_network_filter_close_with_details(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_network_connection_close_type close_type,
+    envoy_dynamic_module_type_buffer_module_ptr details, size_t details_length) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  if (details != nullptr && details_length > 0) {
+    absl::string_view details_view(details, details_length);
+    filter->connection().streamInfo().setConnectionTerminationDetails(details_view);
+  }
+  filter->close(toEnvoyCloseType(close_type));
+}
+
+size_t envoy_dynamic_module_callback_network_filter_get_requested_server_name(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_envoy_ptr* result_out) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  absl::string_view sni = filter->connection().requestedServerName();
+  *result_out = const_cast<char*>(sni.data());
+  return sni.size();
+}
+
+size_t envoy_dynamic_module_callback_network_filter_get_direct_remote_address(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_envoy_ptr* address_out, uint32_t* port_out) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  const auto& address =
+      filter->connection().streamInfo().downstreamAddressProvider().directRemoteAddress();
+
+  if (address == nullptr || address->ip() == nullptr) {
+    *address_out = nullptr;
+    *port_out = 0;
+    return 0;
+  }
+
+  const std::string& addr_str = address->ip()->addressAsString();
+  *address_out = const_cast<char*>(addr_str.c_str());
+  *port_out = address->ip()->port();
+  return addr_str.size();
+}
+
+bool envoy_dynamic_module_callback_network_filter_get_ssl_uri_sans(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_envoy_buffer** sans_out, size_t* sans_count_out) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  const auto ssl = filter->connection().ssl();
+  if (!ssl) {
+    *sans_out = nullptr;
+    *sans_count_out = 0;
+    return false;
+  }
+
+  const auto& uri_sans = ssl->uriSanPeerCertificate();
+  if (uri_sans.empty()) {
+    *sans_out = nullptr;
+    *sans_count_out = 0;
+    return true;
+  }
+
+  // Allocate static thread-local storage for the result.
+  static thread_local std::vector<envoy_dynamic_module_type_envoy_buffer> san_buffers;
+  san_buffers.clear();
+  san_buffers.reserve(uri_sans.size());
```

Comment:
This is great idea to avoid dynamic allocation. But, but also very fragile. Users are required to be very sensitive to the lifetime.

Although I think this could be an option for performance optimization in the future, but for now, could we use our current solution? (module will get size first and then allocate a buffer, then the envoy populate the buffer)